tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,227 @@
+
+/*
+ * sendmsg and recvmsg are syscalls which can be used to send 'real' data along
+ * with ancillary data. Ancillary data is used to communicate to the socket the
+ * type of the TLS record being sent/received.
+ *
+ * Ancillary data macros (CMSG_*) are platform specific and gated.
+ */
+
+S2N_RESULT s2n_ktls_send_msg_impl(int sock, struct msghdr *msg,
+        struct iovec *msg_iov, size_t count, s2n_blocked_status *blocked, ssize_t *send_len)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(send_len);
+    RESULT_ENSURE_GT(count, 0);
+
+    /* set send buffer */
+    msg->msg_iov = msg_iov;
+    msg->msg_iovlen = count;
+
+    *blocked = S2N_BLOCKED_ON_WRITE;
+    *send_len = sendmsg(sock, msg, 0);
+    if (*send_len < 0) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        }
+        RESULT_BAIL(S2N_ERR_IO);
+    }
+    *blocked = S2N_NOT_BLOCKED;
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_set_ancillary_data(struct msghdr *msg, uint8_t record_type)
+{
+    RESULT_ENSURE_REF(msg);
+
+#if !S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+    RESULT_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+#else
+    RESULT_ENSURE_REF(msg->msg_control);
+    RESULT_ENSURE_GTE(msg->msg_controllen, CMSG_SPACE(sizeof(uint8_t)));
+
+    /* set ancillary data */
+    struct cmsghdr *hdr = CMSG_FIRSTHDR(msg);
+    RESULT_ENSURE(hdr != NULL, S2N_ERR_IO);
+    hdr->cmsg_level = S2N_SOL_TLS;
+    hdr->cmsg_type = S2N_TLS_SET_RECORD_TYPE;
+    hdr->cmsg_len = CMSG_LEN(sizeof(uint8_t));
+    RESULT_CHECKED_MEMCPY(CMSG_DATA(hdr), &record_type, sizeof(uint8_t));
+#endif
+
+    return S2N_RESULT_OK;
+}
+
+/* Best practices taken from
+ * https://man7.org/tlpi/code/online/dist/sockets/scm_cred_send.c.html */
+S2N_RESULT s2n_ktls_send_msg(
+        int sock, uint8_t record_type, struct iovec *msg_iov,
+        size_t count, s2n_blocked_status *blocked, ssize_t *send_len)
+{
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(msg_iov->iov_base);
+    RESULT_ENSURE_GT(msg_iov->iov_len, 0);
+    RESULT_ENSURE_GT(count, 0);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(send_len);
+
+    /* Init msghdr
+     *
+     * The control message buffer must be zero-initialized in order for the
+     * CMSG_NXTHDR() macro to work correctly. However since we dont use
+     * CMSG_NXTHDR for kTLS and this code path is performance sensitive we skip
+     * this step.
+     */
+    struct msghdr msg = { 0 };
+
+#if S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+    /* Allocate a char array of suitable size to hold the ancillary data.
+     * However, since this buffer is in reality a 'struct cmsghdr', use a
+     * union to ensure that it is aligned as required for that structure.
+     */
+    union {
+        char buf[CMSG_SPACE(sizeof(uint8_t))];
+        struct cmsghdr _align;
```

Comment:
To understand why alignment is necessary we need to look at `cmsghdr` (ancillary data). The struct defines a length `cmsg_len` but then captures the data, cmsg_data[], as a [flexible array](https://www.ibm.com/docs/en/i/7.2?topic=declarations-flexible-array-members), which allow you to capture data of arbitrary size. However since the underlying struct is still a `cmsghdr`, we need to use a `union` to ensure that the overall allocation aligns with `cmsghdr`.

```
 struct cmsghdr {
     size_t cmsg_len;    /* Data byte count, including header */
     int    cmsg_level;  /* Originating protocol */
     int    cmsg_type;   /* Protocol-specific type */
     unsigned char cmsg_data[]; */
 };
```

p.s. alignment using `union` is also included in the man pages for cmsg https://man7.org/linux/man-pages/man3/cmsg.3.html